crates/wasmtime/src/config.rs

```diff
@@ -292,12 +379,102 @@ pub struct Config {
+    ///
+    /// In general it's encouraged that the integration with `async` and
+    /// wasmtime is designed early on in your embedding of Wasmtime to ensure
+    /// that it's planned that WebAssembly executes in the right context of your
+    /// application.
+    ///
+    /// # Execution in `poll`
+    ///
+    /// The [`Future::poll`](std::future::Future::poll) method is the main
+    /// driving force behind Rust's futures. That method's own documentation
+    /// states "an implementation of `poll` should strive to return quickly, and
+    /// should not block". This, however, can be at odds with executing
+    /// WebAssembly code as part of the `poll` method itself. If your
+    /// WebAssembly is untrusted then this could allow the `poll` method to take
+    /// arbitrarily long in the worst case, likely blocking all other
+    /// asynchronous tasks.
+    ///
+    /// To remedy this situation you have a two possible ways to solve this:
+    ///
+    /// * First you can spawn futures into a thread pool. Care must be taken for
+    ///   this because Wasmtime futures are not `Send` or `Sync`. If you ensure
+    ///   that the entire state of a `Store` is wrapped up in a single future,
+    ///   though, you can send the whole future at once to a separate thread. By
+    ///   doing this in a thread pool you are relaxing the requirement that
+    ///   `Future::poll` must be fast because your future is executing on a
+    ///   separate thread. This strategy, however, would likely still require
+    ///   some form of cancellation via [`crate::Store::interrupt_handle`] to ensure
+    ///   wasm doesn't take *too* long to execute.
+    ///
+    /// * Alternatively you can enable the
+    ///   [`Config::consume_fuel`](crate::Config::consume_fuel) method as well
+    ///   as [`crate::Store::out_of_fuel_async_yield`] When doing so this will
+    ///   configure Wasmtime futures to yield periodically while they're
+    ///   executing WebAssembly code. After consuming the specified amount of
+    ///   fuel wasm futures will return `Poll::Pending` from their `poll`
+    ///   method, and will get automatically re-polled later. This enables the
+    ///   `Future::poll` method to take roughly a fixed amount of time since
+    ///   fuel is guaranteed to get consumed while wasm is executing. Note that
+    ///   to prevent infinite execution of wasm you'll still need to use
+    ///   [`crate::Store::interrupt_handle`].
+    ///
+    /// In either case special care needs to be taken when integrating
+    /// asynchronous wasm into your application. You should carefully plan where
+    /// WebAssembly will execute and what compute resources will be allotted to
+    /// it. If Wasmtime doesn't support exactly what you'd like just yet, please
+    /// feel free to open an issue!
+    #[cfg(feature = "async")]
+    #[cfg_attr(nightlydoc, doc(cfg(feature = "async")))]
+    pub fn new_async() -> Self {
```

Comment:
Hm although then you'd have to also validate that no functions were defined yet, which is a bit wonky...

Do you have thoughts on this? I think it's fine either way myself.

Reply:
I'm leaning towards making it a setting like any other, rather than having `Config::new_async`, and having to set "async" prior to defining any async shared host functions.

With the instance allocator and now host functions in `Config`, I'm concerned that `Config` is too stateful now.  There's a similar problem with `async_stack_size` not being able to be set *after* an instance allocation strategy was set because the pooling allocator relies on that setting.

I'm wondering if we should move the *instantiated* instance allocator and shared host functions into `Engine` (like we have for `Compiler`), leaving `Config` a bunch of settings and only concerned with conflicts rather than *ordering* of the settings.  This wouldn't be a breaking change to the API.

This would change the semantics a little as each engine would now have its own instance allocator, but to me that makes more sense than having each engine share the instance allocator from a "configuration".

We'd still have to panic when you attempt to construct an `Engine` from an incorrect `Config`, though (e.g. you defined an async host function without setting the "async" setting, but now the order of those operations doesn't matter).

What's your take?

Reply:
However, we'd also have to panic when constructing an engine if we can't allocate enough memory for the pooling instance allocator (when used) instead of what we do currently which is return error from `Config::with_allocation_strategy`.

Given this would ever likely be done once early on and isn't recoverable in a way that would make progress (out of memory 🤷), I'm not terribly concerned with a panicking if it improves `Config`.

Reply:
I've pushed up a change for the instance allocator to get a sense of what I'm talking about above. It also removes that confusing "default" instance allocator in the process.

I'll see about the same for moving "async" to a setting and making instantiating the host functions into the engine rather than config.

If we don't like it this way, I'll remove the commits.